java-checks/src/main/java/org/sonar/java/checks/NullPointerCheck.java

```diff
@@ -269,34 +308,105 @@ private void checkForIssue(Tree tree) {
+    if (currentConditionalState != null) {
+      currentConditionalState.mergeConditionalAnd(leftConditionalState, rightConditionalState);
+    }
+  }
+
+  private void visitConditionalOr(BinaryExpressionTree tree) {
+    ConditionalState leftConditionalState = visitCondition(tree.leftOperand());
+    // in case of a conditional or, the current state for the right operand is the false state of the left one,
+    // because of the right operand is evaluated only if the left operand was false.
+    ConditionalState rightConditionalState = visitCondition(tree.rightOperand(), leftConditionalState.falseState);
+    if (currentConditionalState != null) {
+      currentConditionalState.mergeConditionalOr(leftConditionalState, rightConditionalState);
+    }
+  }
+
+  // extracts the symbol in case of ident <op> null, or null <op> ident.
+  @Nullable
+  private VariableSymbol extractRelationalSymbol(BinaryExpressionTree tree) {
+    if (tree.leftOperand().is(Tree.Kind.NULL_LITERAL) && tree.rightOperand().is(Tree.Kind.IDENTIFIER)) {
+      return (VariableSymbol) semanticModel.getReference((IdentifierTreeImpl) tree.rightOperand());
+    } else if (tree.leftOperand().is(Tree.Kind.IDENTIFIER) && tree.rightOperand().is(Tree.Kind.NULL_LITERAL)) {
+      return (VariableSymbol) semanticModel.getReference((IdentifierTreeImpl) tree.leftOperand());
+    } else {
+      return null;
+    }
+  }
+
+  private void visitRelationalEqualTo(BinaryExpressionTree tree) {
+    VariableSymbol symbol = extractRelationalSymbol(tree);
+    if (symbol != null && currentConditionalState != null) {
+      currentConditionalState.trueState.setVariableValue(symbol, AbstractValue.NULL);
+      currentConditionalState.falseState.setVariableValue(symbol, AbstractValue.NOTNULL);
+    }
+  }
+
+  private void visitRelationalNotEqualTo(BinaryExpressionTree tree) {
+    VariableSymbol symbol = extractRelationalSymbol(tree);
+    if (symbol != null && currentConditionalState != null) {
+      currentConditionalState.trueState.setVariableValue(symbol, AbstractValue.NOTNULL);
+      currentConditionalState.falseState.setVariableValue(symbol, AbstractValue.NULL);
+    }
+  }
+
+  private static class ConditionalState {
+    final State falseState;
+    final State trueState;
+
+    ConditionalState(State currentState) {
+      falseState = new State(currentState);
+      trueState = new State(currentState);
+    }
+
+    void mergeConditionalAnd(ConditionalState leftConditionalState, ConditionalState rightConditionalState) {
+      // merges the information for the false state.
+      // the false state of the right operand also contains the information from the true state of the left operand.
+      // first, discards the information learned from the left operand.
+      leftConditionalState.falseState.merge(leftConditionalState.trueState, null);
+      leftConditionalState.falseState.merge(rightConditionalState.falseState, null);
```

Comment:
I am really doubtful about this "merge" in fact. 

in the case of something like : 
```java
if(a !=null && b!=null) {
//...
}  else { 
   a.foo(); //we don't raise issue here because we cancel constraints on "a" (which is fine)
   b.foo(); //shall we raise an issue here ? we have no idea which path  was taken and so it can b that b is perfectly fine, case is not different from a.
}
```